src/bijectors/corr.jl

```diff
@@ -333,47 +339,121 @@ _link_chol_lkj_from_lower(W::AbstractMatrix) = _link_chol_lkj_from_upper(transpo
+    logJ = zero(T)
 
+    idx = 1
     @inbounds for j in 1:K
-        W[1, j] = 1
-        for i in 2:j
-            z = tanh(Y[i - 1, j])
-            tmp = W[i - 1, j]
-            W[i - 1, j] = z * tmp
-            W[i, j] = tmp * sqrt(1 - z^2)
+        log_remainder = zero(T)  # log of proportion of unit vector remaining
+        for i in 1:(j - 1)
+            z = tanh(Y[i, j])
+            idx += 1
+            W[i, j] = z * exp(log_remainder)
+            log_remainder += log1p(-z^2) / 2
+            logJ += log_remainder
         end
+        logJ += log_remainder
+        W[j, j] = exp(log_remainder)
         for i in (j + 1):K
             W[i, j] = 0
         end
     end
 
-    return W
+    return W, logJ
 end
 
 function _inv_link_chol_lkj(y::AbstractVector)
+    LinearAlgebra.require_one_based_indexing(y)
     K = _triu1_dim_from_length(length(y))
 
     W = similar(y, K, K)
+    T = typeof(log(one(eltype(W))))
+    logJ = zero(T)
 
     idx = 1
     @inbounds for j in 1:K
-        W[1, j] = 1
-        for i in 2:j
+        log_remainder = zero(T)  # log of proportion of unit vector remaining
+        for i in 1:(j - 1)
             z = tanh(y[idx])
             idx += 1
-            tmp = W[i - 1, j]
-            W[i - 1, j] = z * tmp
-            W[i, j] = tmp * sqrt(1 - z^2)
+            W[i, j] = z * exp(log_remainder)
+            log_remainder += log1p(-z^2) / 2
+            logJ += log_remainder
+        end
+        logJ += log_remainder
+        W[j, j] = exp(log_remainder)
+        for i in (j + 1):K
+            W[i, j] = 0
+        end
+    end
+
+    return W, logJ
+end
+
+# shared reverse-mode AD rule code
+function _inv_link_chol_lkj_rrule(y::AbstractVector)
+    LinearAlgebra.require_one_based_indexing(y)
+    K = _triu1_dim_from_length(length(y))
+
+    W = similar(y, K, K)
+    T = typeof(log(one(eltype(W))))
+    logJ = zero(T)
+
+    z_vec = tanh.(y)
+
+    idx = 1
+    W[1, 1] = 1
+    @inbounds for j in 2:K
+        log_remainder = zero(T)  # log of proportion of unit vector remaining
+        for i in 1:(j - 1)
+            z = z_vec[idx]
+            idx += 1
+            W[i, j] = z * exp(log_remainder)
+            log_remainder += log1p(-z^2) / 2
+            logJ += log_remainder
         end
+        logJ += log_remainder
+        W[j, j] = exp(log_remainder)
         for i in (j + 1):K
             W[i, j] = 0
         end
     end
 
-    return W
+    function pullback_inv_link_chol_lkj((ΔW, ΔlogJ))
+        LinearAlgebra.require_one_based_indexing(ΔW)
+        Δy = similar(y)
+
+        idx = lastindex(y)
+        @inbounds for j in K:-1:2
+            Δlog_remainder = W[j, j] * ΔW[j, j] + 2ΔlogJ
+            for i in (j - 1):-1:1
+                W_ΔW = W[i, j] * ΔW[i, j]
+                z = z_vec[idx]
+                Δy[idx] = (inv(z) - z) * W_ΔW - z * Δlog_remainder
+                idx -= 1
+                Δlog_remainder += ΔlogJ + W_ΔW
+            end
+        end
```

Comment:
Replacing `idx` with `idx_local` fixes the type-stability issue o.O

```suggestion
    function pullback_inv_link_chol_lkj((ΔW, ΔlogJ))
        LinearAlgebra.require_one_based_indexing(ΔW)
        Δy = similar(y)

        idx_local = lastindex(y)
        @inbounds for j in K:-1:2
            Δlog_remainder = W[j, j] * ΔW[j, j] + 2ΔlogJ
            for i in (j - 1):-1:1
                W_ΔW = W[i, j] * ΔW[i, j]
                z = z_vec[idx_local]
                Δy[idx_local] = (inv(z) - z) * W_ΔW - z * Δlog_remainder
                idx_local -= 1
                Δlog_remainder += ΔlogJ + W_ΔW
            end
        end
```

Reply:
In hindsight, I guess it's not that confusing. From https://docs.julialang.org/en/v1/manual/variables-and-scoping/#local-scope (emphasis mine)

> The next clear cut situation we'll consider is when there is already a local variable named x, in which case x = <value> **always assigns to this existing local x**. This is true whether the assignment occurs in the same local scope, an inner local scope in the same function body, or in the body of a function nested inside of another function, also known as a [closure](https://en.wikipedia.org/wiki/Closure_(computer_programming)).

Reply:
Makes sense when you think about scenarios such as:
```julia
function f(x)
    function g(y)
        x = 1
        return y + x
    end

    return g, y -> y + x
end

g, h = f(3)
h(1) # => 4
g(1) # => 2
h(1) # => 2 (because `x = 1` when we call `g`)
```